src/nvim/edit.c

```diff
@@ -3825,6 +3831,21 @@ static void ins_compl_insert(void)
     compl_used_match = FALSE;
   else
     compl_used_match = TRUE;
+
+  // Set completed item.
+  // { word, abbr, menu, kind, info }
+  dict_T *dict = dict_alloc();
+  dict_add_nr_str(dict, "word", 0L,
+                  SHOULD_BE_STR(compl_shown_match->cp_str));
+  dict_add_nr_str(dict, "abbr", 0L,
+                  SHOULD_BE_STR(compl_shown_match->cp_text[CPT_ABBR]));
+  dict_add_nr_str(dict, "menu", 0L,
+                  SHOULD_BE_STR(compl_shown_match->cp_text[CPT_MENU]));
+  dict_add_nr_str(dict, "kind", 0L,
+                  SHOULD_BE_STR(compl_shown_match->cp_text[CPT_KIND]));
+  dict_add_nr_str(dict, "info", 0L,
+                  SHOULD_BE_STR(compl_shown_match->cp_text[CPT_INFO]));
```

Comment:
Did you set it on the `dictitem_S` for each value? For example, `v:cmdarg` is readonly. `v:` is a dictionary, and `cmdarg` is the value. It should be possible to make `v:foo.bar` readonly also.

It may require some digging into `dict_add_nr_str()`. Inside that function, `item->di_tv.di_flags` should be set.